File: app/src/main/java/bassamalim/hidaya/activities/BookSearcher.java
package bassamalim.hidaya.activities;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.SearchView;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import bassamalim.hidaya.R;
import bassamalim.hidaya.adapters.BookSearcherAdapter;
import bassamalim.hidaya.databinding.ActivityBookSearcherBinding;
import bassamalim.hidaya.models.Book;
import bassamalim.hidaya.models.BookSearcherMatch;
import bassamalim.hidaya.other.Utils;
import bassamalim.hidaya.dialogs.FilterDialog;

public class BookSearcher extends AppCompatActivity {

    private ActivityBookSearcherBinding binding;
    private SharedPreferences pref;
    private Gson gson;
    private String[] bookTitles;
    private List<BookSearcherMatch> matches;
    private RecyclerView recyclerView;
    private SearchView searchView;
    private BookSearcherAdapter adapter;
    private int maxMatches;
    private boolean[] selectedBooks;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        binding = ActivityBookSearcherBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        init();

        setListeners();

        initFilterIb();

        setupSizeSpinner();
    }

    private void init() {
        pref = PreferenceManager.getDefaultSharedPreferences(this);

        gson = new Gson();

        bookTitles = getResources().getStringArray(R.array.books_titles);

        searchView = binding.searchView;

        matches = new ArrayList<>();

        maxMatches = pref.getInt("books_searcher_matches_last_position", 10);
    }

    private void setListeners() {
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                perform(query);
                return true;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                return true;
            }
        });

        binding.filterIb.setOnClickListener(v ->
                new FilterDialog<>(this, v, "اختر الكتب", bookTitles, selectedBooks,
                        adapter, binding.filterIb, "selected_search_books"));
    }

    private void setupSizeSpinner() {
        Spinner spinner = binding.sizeSpinner;
        int last = pref.getInt("books_searcher_matches_last_position", 0);
        spinner.setSelection(last);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long vId) {
                maxMatches = Integer.parseInt(spinner.getItemAtPosition(position).toString());

                SharedPreferences.Editor editor = pref.edit();
                editor.putInt("books_searcher_matches_last_position", position);
                editor.apply();

                if (adapter != null && adapter.getItemCount() > 0)
                    perform(searchView.getQuery().toString());
            }
            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void perform(String query) {
        selectedBooks = getSelectedBooks();
        search(query);

        if (matches.isEmpty()) {
            binding.notFoundTv.setVisibility(View.VISIBLE);
            binding.recycler.setVisibility(View.INVISIBLE);
        }
        else {
            binding.notFoundTv.setVisibility(View.INVISIBLE);
            binding.recycler.setVisibility(View.VISIBLE);

            setupRecycler(matches);

            if (adapter != null)
                adapter.notifyDataSetChanged();
        }
    }

    private void search(String text) {
        matches.clear();

        String prefix = "/Books/";
        File dir = new File(getExternalFilesDir(null) + prefix);

        if (!dir.exists())
            return;

        File[] files = dir.listFiles();

        for (int i = 0; i < Objects.requireNonNull(files).length; i++) {
            if (!selectedBooks[i])
                continue;

            String jsonStr = Utils.getJsonFromDownloads(getExternalFilesDir(null) +
                    prefix + i + ".json");
            Book book = gson.fromJson(jsonStr, Book.class);

            for (int j = 0; j < book.getChapters().length; j++) {
                Book.BookChapter chapter = book.getChapters()[j];

                for (int k = 0; k < chapter.getDoors().length; k++) {
                    Book.BookChapter.BookDoor door = chapter.getDoors()[k];
                    String doorText = door.getText();

                    Matcher m = Pattern.compile(text).matcher(doorText);
                    Spannable ss = new SpannableString(doorText);
                    while (m.find()) {
                        ss.setSpan(new ForegroundColorSpan(getColor(R.color.highlight_M)),
                                m.start(), m.end(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

                        matches.add(new BookSearcherMatch(
                                i, book.getBookInfo().getBookTitle(),
                                j, chapter.getChapterTitle(),
                                k, door.getDoorTitle(), ss));

                        if (matches.size() == maxMatches)
                            return;
                    }
                }
            }
        }
    }

    private void setupRecycler(List<BookSearcherMatch> matches) {
        recyclerView = binding.recycler;
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new BookSearcherAdapter(matches, searchView);
        recyclerView.setAdapter(adapter);
    }

    private void initFilterIb() {
        selectedBooks = getSelectedBooks();
        for (boolean bool : selectedBooks) {
            if (!bool) {
                binding.filterIb.setImageDrawable(
                        AppCompatResources.getDrawable(this, R.drawable.ic_filtered));
                break;
            }
        }
    }

    private boolean[] getSelectedBooks() {
        boolean[] defArr = new boolean[bookTitles.length];
        Arrays.fill(defArr, true);
        String defStr = gson.toJson(defArr);
        return gson.fromJson(pref.getString("selected_search_books", defStr), boolean[].class);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
        if (recyclerView != null)
            recyclerView.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/BookSearcher.kt
package bassamalim.hidaya.activities

import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.preference.PreferenceManager
import bassamalim.hidaya.R
import bassamalim.hidaya.database.AppDatabase
import bassamalim.hidaya.database.dbs.BooksDB
import bassamalim.hidaya.dialogs.FilterDialog
import bassamalim.hidaya.models.Book
import bassamalim.hidaya.models.BookSearcherMatch
import bassamalim.hidaya.other.Global
import bassamalim.hidaya.ui.components.*
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import bassamalim.hidaya.utils.FileUtils
import bassamalim.hidaya.utils.PrefUtils
import com.google.gson.Gson
import java.io.File
import java.util.regex.Pattern

class BookSearcher : ComponentActivity() {

    private lateinit var pref: SharedPreferences
    private lateinit var db: AppDatabase
    private val gson = Gson()
    private lateinit var books: List<BooksDB>
    private lateinit var language: String
    private val selectedBooks = mutableStateListOf<Boolean>()
    private val filteredState = mutableStateOf(false)
    private val maxMatchesIndex = mutableStateOf(0)
    private lateinit var maxMatchesItems: Array<String>
    private val matches = mutableStateListOf<BookSearcherMatch>()
    private var searched = false
    private val filterDialogShown = mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        language = ActivityUtils.myOnActivityCreated(this)[1]

        init()

        setContent {
            AppTheme {
                UI()
            }
        }
    }

    private fun init() {
        pref = PreferenceManager.getDefaultSharedPreferences(this)
        db = DBUtils.getDB(this)

        books = db.booksDao().getAll()

        initSelectedBooks()
        initFilterIb()
        initMaxMatches()
    }

    private fun initSelectedBooks() {
        for (i in books.indices) selectedBooks.add(true)
        val json = PrefUtils.getString(pref, "selected_search_books", "")
        if (json.isNotEmpty()) {
            val boolArr =  gson.fromJson(json, BooleanArray::class.java)
            for (i in boolArr.indices) selectedBooks[i] = boolArr[i]
        }
    }

    private fun initFilterIb() {
        for (bool in selectedBooks) {
            if (!bool) {
                filteredState.value = true
                break
            }
        }
    }

    private fun initMaxMatches() {
        maxMatchesIndex.value = PrefUtils.getInt(
            pref, "books_searcher_max_matches_index", 0
        )

        maxMatchesItems =
            if (language == "en") resources.getStringArray(R.array.searcher_matches_en)
            else resources.getStringArray(R.array.searcher_matches)
    }

    private fun search(text: String, highlightColor: Color) {
        matches.clear()

        val prefix = "/Books/"
        val dir = File(getExternalFilesDir(null).toString() + prefix)
        if (!dir.exists()) return

        for (i in books.indices) {
            if (!selectedBooks[i] || !downloaded(i)) continue

            val jsonStr = FileUtils.getJsonFromDownloads(
                getExternalFilesDir(null).toString() + prefix + i + ".json"
            )
            val book =
                try {
                    gson.fromJson(jsonStr, Book::class.java)
                } catch (e: Exception) {
                    Log.e(Global.TAG, "Error in json read in BookSearcher")
                    e.printStackTrace()
                    continue
                }

            for (j in book.chapters.indices) {
                val chapter = book.chapters[j]

                for (k in chapter.doors.indices) {
                    val door = chapter.doors[k]
                    val doorText = door.text

                    val matcher = Pattern.compile(text).matcher(doorText)
                    if (matcher.find()) {
                        val annotatedString = buildAnnotatedString {
                            append(doorText)

                            do {
                                addStyle(
                                    style = SpanStyle(color = highlightColor),
                                    start = matcher.start(),
                                    end = matcher.end()
                                )
                            } while (matcher.find())
                        }

                        matches.add(
                            BookSearcherMatch(
                                i, book.bookInfo.bookTitle,
                                j, chapter.chapterTitle,
                                k, door.doorTitle, annotatedString
                            )
                        )

                        searched = true
                        if (matches.size == maxMatchesItems[maxMatchesIndex.value].toInt()) return
                    }
                }
            }
        }
    }

    private fun downloaded(id: Int): Boolean {
        val dir = File(getExternalFilesDir(null).toString() + "/Books/")

        if (!dir.exists()) return false

        val files = dir.listFiles()
        for (element in files!!) {
            val name = element.name
            val n = name.substring(0, name.length - 5)
            try {
                val num = n.toInt()
                if (num == id) return true
            } catch (ignored: NumberFormatException) {}
        }
        return false
    }

    @Composable
    private fun UI() {
        MyScaffold(stringResource(R.string.books_searcher)) { padding ->
            val highlightColor = AppTheme.colors.accent
            val textState = remember { mutableStateOf(TextFieldValue("")) }

            Column(
                Modifier
                    .fillMaxSize()
                    .padding(padding),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Column(
                    Modifier.fillMaxWidth(),
                    verticalArrangement = Arrangement.SpaceAround,
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    MyText(
                        text = stringResource(R.string.search_in_books),
                        modifier = Modifier.padding(vertical = 6.dp)
                    )

                    SearchComponent(
                        state = textState,
                        hint = stringResource(R.string.search),
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 30.dp),
                        onSubmit = {
                            search(
                                text = textState.value.text,
                                highlightColor = highlightColor
                            )
                        }
                    )

                    Row(
                        Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceEvenly
                    ) {
                        MyText(text = stringResource(R.string.selected_books))

                        MyIconBtn(
                            iconId = R.drawable.ic_filter,
                            description = stringResource(R.string.filter_search_description),
                            size = 30.dp,
                            tint =
                                if (filteredState.value) AppTheme.colors.secondary
                                else AppTheme.colors.weakText
                        ) {
                            filterDialogShown.value = true
                        }
                    }

                    Row(
                        Modifier
                            .fillMaxWidth()
                            .padding(bottom = 5.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceEvenly
                    ) {
                        MyText(text = stringResource(R.string.max_num_of_marches))

                        MyDropDownMenu(
                            selectedIndex = maxMatchesIndex,
                            items = maxMatchesItems
                        ) { index ->
                            pref.edit()
                                .putInt("books_searcher_max_matches_index", index)
                                .apply()
                        }
                    }
                }

                if (matches.isEmpty()) {
                    if (searched)
                        MyText(
                            text = stringResource(R.string.books_no_matches),
                            modifier = Modifier.padding(top = 100.dp)
                        )
                }
                else {
                    MyLazyColumn(lazyList = {
                        items(
                            items = matches
                        ) { item ->
                            MySurface {
                                Column(
                                    horizontalAlignment = Alignment.CenterHorizontally
                                ) {
                                    MyText(
                                        text = item.bookTitle,
                                        fontWeight = FontWeight.Bold,
                                        modifier = Modifier.padding(6.dp)
                                    )
                                    MyText(
                                        text = item.chapterTitle,
                                        modifier = Modifier.padding(6.dp)
                                    )
                                    MyText(
                                        text = item.doorTitle,
                                        modifier = Modifier.padding(6.dp)
                                    )
                                    MyText(
                                        text = item.text,
                                        modifier = Modifier.padding(6.dp)
                                    )
                                }
                            }
                        }
                    })
                }
            }

            if (filterDialogShown.value) {
                val bookTitles =
                    if (language == "en") db.booksDao().getTitlesEn()
                    else db.booksDao().getTitles()
                FilterDialog(
                    pref, gson, getString(R.string.choose_books),
                    bookTitles, selectedBooks, filteredState,
                    { if (searched) search(textState.value.text, highlightColor) },
                    "selected_search_books", filterDialogShown
                ).Dialog()
            }
        }
    }
}